Develop

CezaryZal opened the pull request:
Add spring security, create POST method to add new user and connect application with HealthCalendar app

CezaryZal committed:
Change implement object to `UserService` interface from implement class

CezaryZal committed:
Update app to use spring security

Move `getRoleList` and `getPermissionList` methods to `UserServer` interface;
Add "/api" path to `LoginController` class and remove `NameNotFoundException` because it is not needed

CezaryZal committed:
Add spring security

Create `SecurityConfig` class;
Add spring-security dependency to pom.xml

CezaryZal committed:
Add password encryption and connect security with db
Create `ContainersCreator` class in manager.db.service package to split String roles/permission on List;
Create `security` package, it contain:
new `Bcrypt` class to encrypt and decryption a password by `BcryptPasswordEncoder()`,
new `UserPrincipal` class to convert the `User` entity on UserDetails,
new `UserPrincipalDetailsService` class to convert the input formLogin using the `UserPrincipal` class;
Remove `getRoleList` and `getPermissionList` methods from `UserServiceImp` class and update `addNewUser` method by adding password encryption;
Update `template_data_user_server.sql` to use password encryption

CezaryZal committed:
Rename entity to `UserToHc` and related method, classes, fields

Rename `UserToHc` from `UserToDb` name;
Change table name in database to `user_to_hc`

CezaryZal committed:
Add new `user_app` table with data and update to use

Rename classes in manager.db.service and controller package to `UserHc...` from `User...`;
Create `UserApp` class for the new entity used for authentication;
Create `UserAppRepository` class;
Update to `template_data_server.sql` name file and data;
Swap resource to `UserApp` entity from `UserHc` entity;

CezaryZal committed:
Rename entity `UserHc` and `ContainersCreatorOfUserHc` class

Name `UserHc` (user to Health Calendar application) from `UserToHc`;
Create modifier.entity package to `ContainersCreatorOfUserHc` class

CezaryZal committed:
Code reduction in `UserApp` class

Remove @NoArgsConstructor annotation and change @Data to @Getter annotation from Lombok

CezaryZal committed:
Add encryption in comparator passwords

Move location `SecurityConfig` and `SecurityConstants` classes to `security` package

CezaryZal committed:
Rename package and remove `SecurityConfig` class

Rename to the `authentication` package from `security` package;
Remove `SecurityConfig` to `config` package from `security` package

CezaryZal committed:
Connection App with HealthCalendar app

Create `app` and `health.calendar` package;
Create `UserCreator` class in `health.calendar` package;
Create `TestController` class to test connection with HealthCalendar application;
Rename some classes and methods;

CezaryZal committed:
Add endpoint in TestController.java

POST method connection with HealthCalendar application;

CezaryZal committed:
Add endpoint "/new-account" in LoginController.java

Rename entity to ConnectingUser.java from UserCreator.java;
Create InputUser.java entity, necessary when creating a new account;
Create `connector` package with AccountCreator.java to sent object for HealthCalendar application and create new account(User);
Create Converter.java to convert InputUser to UserAuthentication;
Add `createNewAccount` method in LoginService.java;

CezaryZal committed:
Add token creator to endpoint "/new-account"

Add lombok library to ConnectingUser.java;
Create UserConstants.java;
Create `constants` package and move there the SecurityConstants.java and UserConstants.java;
Create AuthRequestByUserConstant.java, HeadersByToken.java, HttpEntityByBodyAndToken.java, NewUser.java classes in `creator` package;
Update TestController.java, create endpoint with used token;

CezaryZal committed:
Add validation InputUser.java and extends superclass entity

Create FormUser.java as superclass for input entity;
Update to use new class of form;

CezaryZal committed:
Add email validation from InputUser entity

Create EmailFormValidator.java;
Create InvalidEmailFormException.java;
Rename main class validation service;

CezaryZal committed:
Add role validation from InputUser entity

Create RoleValidator.java;
Create IncorrectRoleException.java;
Add new exception class to RestResponseEntityExceptionHandler.java;
Rename method in FormValidator.java to `isNull`

CezaryZal committed:
Add login name validation, check exists in DB

Create LoginNameComparator.java;
Create LoginNameExistsException.java;
Create endpoint in UserHcAuthRepository.java ("/exist") by login name nad contains method in HcRepository;
Add new exception class to RestResponseEntityExceptionHandler.java;

s235jr commented on src/main/java/com/CezaryZal/authentication/Bcrypt.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Bean
+    PasswordEncoder passwordEncoder(){
+        return new BCryptPasswordEncoder();
```
To możesz sobie wrzucić w SecurityConfig, wtedy nie potrzebujesz tej klasy.

CezaryZal replied:
Ok, przeniosłem.
Myślałem żeby zrobić oddzielną klasę na tego beana ponieważ korzystam z niego nie tylko w klasie SecurotyConfig.

s235jr commented on src/main/java/com/CezaryZal/authentication/UserAppPrincipal.java:
```diff
@@ -0,0 +1,65 @@
+
+public class UserAppPrincipal implements UserDetails {
+
+    private UserApp userApp;
```
final?

CezaryZal replied:
Ok

s235jr commented on src/main/java/com/CezaryZal/authentication/UserAppPrincipal.java:
```diff
@@ -0,0 +1,65 @@
+        // Extract list of roles (ROLE_name)
+        String roles = userApp.getRoles();
+        containersCreatorOfUserHcAuth.getRoleList(roles).forEach(r -> {
+            GrantedAuthority authority = new SimpleGrantedAuthority("ROLE_" + r);
```
Coś kojarze że ten prefix "ROLE_" jest dodawany przez spring security, Upewnij się czy Twoje role nie są finalnie w stylu ROLE_ROLE_ADMIN

CezaryZal replied:
Niestety nie, przy budowie tokenu też trzeba dodać taki prefix.
Może Ty korzystasz z innej matody/klasy, która sama go dopisuje?

s235jr commented on src/main/java/com/CezaryZal/authentication/UserAppPrincipalDetailsService.java:
```diff
@@ -0,0 +1,29 @@
+@Service
+public class UserAppPrincipalDetailsService implements UserDetailsService {
+
+    private UserAppRepository userAppRepository;
```
final?

CezaryZal replied:
Ok

s235jr commented on src/main/java/com/CezaryZal/authentication/UserAppPrincipal.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Collection;
+import java.util.List;
+
+public class UserAppPrincipal implements UserDetails {
```
nie da się tutaj wykorzystać package org.springframework.security.core.userdetails.User ? Potrzebujesz włąsnej implementacji?

CezaryZal replied:
Patrząc w dokumentację, pewnie się da. Narazie zrobiłem sposobem jaki znam

s235jr commented on src/main/java/com/CezaryZal/config/SecurityConfig.java:
```diff
@@ -0,0 +1,56 @@
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
```
modyfikatory

s235jr commented on src/main/java/com/CezaryZal/constants/UserConstants.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserConstants {
```
Co to? po co to?

CezaryZal replied:
Obecnie żebym mógł stworzyć nowe konto muszę najpierw stworzyć nowego użytkownika w aplikacji HealthCalendar, później wraz z jego id stworzyć użytkownika w tej aplikacji.
Żeby stworzyć użytkownika w HC potrzebuje tokenu, a żeby stworzyć token muszę najpierw się uwierzytelnić więc wykorzystuje dane istniejącego użytownika do tego.
Jego dane  przechowuje w zewnętrznym pliku z którego pobieram dane przez tą klasę.
Później będę chciał to zrobić przez pominięcie uwierzytelienia, tylko odrazu stworzenia tokenu.

s235jr replied:
Totalnie nie rozumiem tego konceptu, ale ok. Natomiast kod jest przez to bardzo nieczytelny.

s235jr commented on src/main/java/com/CezaryZal/controller/LoginController.java:
```diff
@@ -23,8 +22,12 @@ public LoginController(LoginService loginService) {
 
     @PostMapping("/login")
-    public ResponseEntity<String> getTokenByUserLogin(@RequestBody AuthenticationRequest inputAuthenticationRequest) throws NameNotFoundException, AccountNotFoundException {
+    public ResponseEntity<String> getTokenByUserLogin(@RequestBody AuthenticationRequest inputAuthenticationRequest) throws AccountNotFoundException {
```
Zmieniłeś exception ( i słusznie) a pojawił siędo niego Handler?

CezaryZal replied:
Jest w klasie LoginService w metodzie `throwExceptionIfUserIsNotActive`

s235jr replied:
Tam go rzucasz, a ja pytam o handler

s235jr commented on src/main/java/com/CezaryZal/controller/UserHcController.java:
```diff
@@ -0,0 +1,24 @@
+    List<UserAuthentication> findAll();
+
+    UserAuthentication addNewUser(UserAuthentication userAuthentication);
+
```
UserAuthentication to encja, raczej nie powinna pojawić się ani w request ani w response

CezaryZal replied:
W requescie, racja ale w response czemu nie?
Tutaj są endpointy dla aministratora żeby można było bezpośrednio komunikować się z repozytorium.

s235jr replied:
Bo jakakolwiek zmiana obiektu encji wymusza zmiane API i w drugą strone, zmiana API wymusza zmianę struktury bazy danych

s235jr replied:
https://devszczepaniak.pl/wstep-do-rest-api/ pkt.4

s235jr commented on src/main/java/com/CezaryZal/controller/UserHcController.java:
```diff
@@ -0,0 +1,24 @@
+
+    ResponseEntity<UserAuthentication> findByLoginName(String loginName) throws NameNotFoundException;
+
+    boolean isExistsUserAuthenticationByLoginName(String loginName);
```
co sprawdza ta metoda? nie rozumiem znaczenia opisu

CezaryZal replied:
czy jest istniejszący UserAuthentication sprawdzając na podstawie loginu.
Możesz zaproponować bardziej zrozumiałą nazwę?

s235jr replied:
isUser? isAccount?

s235jr commented on src/main/java/com/CezaryZal/controller/UserHcController.java:
```diff
@@ -0,0 +1,24 @@
+
+    ResponseEntity<UserAuthentication> findById(Long id) throws NameNotFoundException;
+
+    ResponseEntity<UserAuthentication> findByLoginName(String loginName) throws NameNotFoundException;
```
Cy tu nie powinien lecieć AccountNotFound?

CezaryZal replied:
Upss, pozostałość po wcześniejszym wyjątku.
Jest niepotrzebne, zostało obsłużone wcześniej.
W implementacji tego nie ma.

s235jr commented on src/main/java/com/CezaryZal/entity/FormUser.java:
```diff
@@ -0,0 +1,19 @@
+
+    private String loginName;
+    private String password;
+    private String roles;
+    private String email;
```
Stworzyłeś pola roles i email, a nigdzie nie są ustawiane. Nie są zbędne?

CezaryZal replied:
Poniżej odpowiedź

s235jr commented on src/main/java/com/CezaryZal/entity/FormUser.java:
```diff
@@ -0,0 +1,19 @@
+
+@Data
+@NoArgsConstructor
+public class FormUser {
```
Po co jest ten obiekt?

s235jr commented on src/main/java/com/CezaryZal/entity/app/AuthenticationRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+import com.CezaryZal.entity.FormUser;
+
+public class AuthenticationRequest extends FormUser {
```
Co daje dziedziczenie po FormUser?

CezaryZal replied:
Jest to nadklasa, która posiada najważniejsze pola przy logowaniu, tworzeniu nowego użytkownika w tej aplikacji i HealthCalendar aplikacji.
W sumie to zamieniłem ją na klasę abstrakcyjną, będzie tak lepiej.

s235jr commented on src/main/java/com/CezaryZal/entity/app/UserApp.java:
```diff
@@ -0,0 +1,19 @@
+    private Long id;
+    private String loginName;
+    private String password;
+    private String roles;
```
User może mieć tylko jedną role? Nie powinien się tu pojawić Set<String> roles?

CezaryZal replied:
Tutaj nie.
Jest encja do uwierzytelenia użytkownika z tą aplikacją.
Nie przewiduje tutaj żadnych dodatkowych uprawnień po za ADMIN i MODERATOR oraz więcej użytkowników niż dwa z tymi uprawnieniami.

s235jr replied:
Mimo wszystko uważam że powinien tu być Set, bazowanie na stringu, cięcie go itp itd... Po co sobie utrudniać?

s235jr commented on src/main/java/com/CezaryZal/entity/health/calendar/ConnectingUser.java:
```diff
@@ -0,0 +1,15 @@
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+public class ConnectingUser {
```
co to za kolejny user?

CezaryZal replied:
Tutaj miałem problem z nazwą.
Jest to encja pozwalająca stworzyć użytkownika w aplikacji HealthCalendar.
Żeby stworzyć użytkownika app HC potrzebuje tylko loginu i emaila.
Taka sama encja znajduje się również w aplikacji HC

s235jr commented on src/main/java/com/CezaryZal/entity/health/calendar/UserAuthentication.java:
```diff
@@ -0,0 +1,25 @@
+    private String loginName;
+    private String password;
+    private boolean approved;
+    private String roles;
```
To samo co wyżej

CezaryZal replied:
Encja połączona z bazą danych, pozwalająca uwierzytelnić wpisywane dane użytkownika oraz stworzyć token

s235jr commented on src/main/java/com/CezaryZal/entity/health/calendar/UserAuthentication.java:
```diff
@@ -0,0 +1,25 @@
+
+import javax.persistence.*;
+
+@Entity
+@Table(name = "users_hc")
+@Data
+@NoArgsConstructor
+public class UserAuthentication {
```
Czemu masz dwie tabele z userami?

CezaryZal replied:
W skrócie.
Jedna tabela uwierzytelnia Klienta bezpośrednio podbijającego do API (frontend).
Druga tabela do uwierzytelnienia wpisywanych danych, czyli klienta frontendu (zwykłego użytkownika aplikacji HealthCalendar)

s235jr commented on src/main/java/com/CezaryZal/exceptions/InvalidEmailFormException.java:
```diff
@@ -0,0 +1,8 @@
+package com.CezaryZal.exceptions;
+
+public class InvalidEmailFormException extends RuntimeException {
```
InvalidEmailException? Co jest niepoprawne? Cały formularz? Czy email?

CezaryZal replied:
Czy jest niepoprawna forma emaila.
Ona sprawdza tyko formą, a nie czy email jet nie poprawny/istnieje.
Jeśli masz propozycję nazwy, śmiało pisz :)

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -15,35 +21,56 @@
+    private HttpEntityByBodyAndToken httpEntityBodyToken;
+    private NewUser newUser;
 
-    @Autowired
```
czemu znikneła adnotacja @Autowired?

CezaryZal replied:
Kurcze, zapomniałem.
Przy tej klasie często zmieniałem wstrzykiwane klasy i widać że przed wrzuceniem na gita zapomniałem...

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -15,35 +21,56 @@
+        Long userIdFromHc = accountCreator.createAccountInHealthCalendarAndGetUserId(httpEntity);
+        newUser.createNewUser(userIdFromHc, inputUser);
+
+        return "Został stworzony nowy użytkownik";
```
Czy odpowiedzialnoścą LoginService powinno być tworzenie usera? Czy nie powinien tego robić jakis UserService albo jakiś AccountService?

CezaryZal replied:
Też o tym myślałem.
Ale wtedy LoginService miałby tylko taką samą ilość metod co endpointów po jednej klasie obsługującej jedną funkcję.
Dlatego ją rozbudowałem, ale pewien tej archotektury nie jest pewien

s235jr commented on src/main/java/com/CezaryZal/manager/connector/AccountCreator.java:
```diff
@@ -0,0 +1,22 @@
+                Long.class);
+        return responseEntity.getBody();
+    }
+}
```
A to to co?

CezaryZal replied:
Klasa tworzy nowe konto, czyli łączy się z główną aplikacją HealtchCalendar, przesyła dane do stworzenia tam użytkownika i otrzywmuje oraz zwraca id nowo utworzonego użytkownika.

s235jr commented on src/main/java/com/CezaryZal/manager/LoginService.java:
```diff
@@ -15,35 +21,56 @@
+
+        String clearToken = getTokenByUserLogin(authRequestByUserConstant.createAuthRequest());
+        HttpEntity<Object> httpEntity = httpEntityBodyToken.createHttpEntity(connectingUser, clearToken);
+        Long userIdFromHc = accountCreator.createAccountInHealthCalendarAndGetUserId(httpEntity);
```
co Tu się dzieje? robisz request do swojego włąsnego api? po co?

CezaryZal replied:
Jak powyżej opisałem.
Robię requesta do aplikacji HC i otrzymuje od niej id nowo utworzonego użytkownika

s235jr replied:
a gdzie jest ta druga aplikacja? gdzie jest jej kod?

s235jr commented on src/main/java/com/CezaryZal/manager/creator/NewUser.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class NewUser {
```
Co to? jakiś kolejny sposób na tworzenie usera?

CezaryZal replied:
Kontynuując powyższe, gdy mamy id użytkownika z app HC, przypisujemy go do obiektu, który zostanie zapisany w db, tabeli user_hc.

s235jr commented on src/main/java/com/CezaryZal/manager/creator/AuthRequestByUserConstant.java:
```diff
@@ -0,0 +1,13 @@
+    public AuthenticationRequest createAuthRequest(){
+        return new AuthenticationRequest(UserConstants.LOGIN, UserConstants.PASSWORD);
+    }
+}
```
Do takiego czegoś nie potrzeba mieszać springa.

s235jr commented on src/main/java/com/CezaryZal/manager/creator/HeadersByToken.java:
```diff
@@ -0,0 +1,17 @@
+
+        return headers;
+    }
+}
```
Myśle że tego nie powinno być bo request do którego tego używasz powinien zniknąć. A natet jeśli to coś takiego należałoby napisać przez wzorzec Factory. Czyli np klase która ma metode statyczną, która zwraca np RestTemplate (lub dowolny inny klient Http), który ma już wpisany ten token (cośw stylu prefabrykatu obiektu resttemplaete).

CezaryZal replied:
Obecnie nie jestem wstanie zrobić metody statycznej która budowałaby token w sobie, ponieważ używam spring boota. W klasie ze statyczną metodą tworzącą requesta nie mogę wstrzyknąć klas sewisowych.
Ale samo tworzenie nagłówka i ciała do requesta wrzuciłem do metody statycznej.

s235jr commented on src/main/java/com/CezaryZal/manager/creator/HttpEntityByBodyAndToken.java:
```diff
@@ -0,0 +1,22 @@
+        MultiValueMap<String, String> headers = headersByToken.createHeadersByToken(clearToken);
+        return new HttpEntity<>(body, headers);
+    }
+}
```
Nazwa klasy totalnie nic mi nie mówi za co jest odpowiedzialna. 
To chyba też powinno wylecieć, bo to do tego requestu

CezaryZal replied:
Obecnie jak powyżej, zrobiłem z niej metodę statyczną

s235jr commented on src/main/java/com/CezaryZal/manager/filters/comparator/PasswordComparator.java:
```diff
@@ -1,13 +1,14 @@
-        if (!passwordFromUserLogin.equals(userPasswordFromDb)){
+    public void throwIfIsNotEqualsPassword(String passwordFromUserLogin, String userPasswordFromDb) {
+        if (!BCrypt.checkpw(passwordFromUserLogin, userPasswordFromDb)){
             throw new IncorrectLoginOrPasswordSecurityException("Login or password incorrectly entered");
```
Logowanie ma kilka etapów.
Przychodzi request z loginem i hasłem.
1. Sprawdzasz czy istnieje login. Jeżeli nie istnieje login to rzucasz exception AccountNotFound
2.Sprawdzasz czy hasło jest dobre. Jeżeli nie to rzucasz exception że InvalidPasswordException
3. Jeżeli któryś z tych exceptionów poleciał, to wyżej (w miejscu validacji czy tego usera możesz zalogować łapiesz go i rzucasz InvalidLoginOrPassword

CezaryZal replied:
Ok, myślałem że skoro i tak na końcu dajemy ogólny wyjątek to nie ma sensu tego rozbijać, ale to przydatne przy czytaniu logów.

CezaryZal replied:
Ok, obsługę pierwszych wyjątków zrobiłem przez @ControllerAdvice w RestResponseEntityExceptionHandler przez wyłapanie i zmianę Stringa w body.
W taki sposób czy chodziło Ci o Try()catch()?

s235jr replied:
Tak, przez try&catch.
Jeżeli złapiesz AccountNotFount i InvalicPassword i w ich miejsce rzucisz InvalidUserNameOrPassword to nie musisz mieć handlerów dla AccountNotFoun i Invalid Password

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/EmailFormValidator.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private boolean isValid(String email) {
+        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$";
+        String regex = "^[\\w-_\\.+]*[\\w-_\\.]\\@([\\w]+\\.)+[\\w]+[\\w]$";
+
+        Pattern pat = Pattern.compile(emailRegex);
```
To można wynieść do pola klasy. Wystarczy że skompiluje sięto raz i póżniej validator będzie z tego korzystał. Taka optymalizacja wydajnościowa.
Do tego regex jest nieużywany.

CezaryZal replied:
Hyyy, ale i tak metoda jak i cała klasa skompiluje się raz.
Skoro tylko w jednej metodzie używamy nie ma sensu przenosić do pola klasy.

Zostawiłem regex bo w nie wiem który jest bardziej powszechny.
Uważasz że który jest bardziej powszechny?

s235jr replied:
Nie, pattern w metodzie będzie się kompilował przy każdym wywołaniu metody. 
Skąd pomysł że miałby się tylk oraz kompilować?

Wybrałbym regexa który jest prawidłowy. To pozostawiam Tobie.

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/FormValidator.java:
```diff
@@ -9,8 +10,12 @@
 
-    boolean isEmpty(AuthenticationRequest authenticationRequest) {
-        return authenticationRequest == null;
+    boolean isNull(FormUser formUser) {
+        return formUser == null;
+    }
+
+    boolean isNull(String arg){
+        return arg == null;
```
Logika tych metod mówi że to raczej powinno być setNull niż isNull

CezaryZal replied:
SetNull?
Tu nic nie przesyłamy, tylko sprawdzamy, czemu set?

s235jr replied:
Racja, nie zauważyłem że są dwa "="

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/RoleValidator.java:
```diff
@@ -0,0 +1,31 @@
+@Component
+public class RoleValidator {
+
+    List<String> listRoles = Arrays.asList("ADMIN", "USER", "VIEWER");
```
Dla takich ról warto stworzyć sobie enum-y. Pewnie tego nie robiłeś, wiec możesz poćwiczyc

CezaryZal replied:
Robiłem i nawet zrobiłem ale stwierdziłem że dla tak prostego zastosowania nie ma sensu.
Ale mogę w taki razie zrobić :)

s235jr commented on src/main/java/com/CezaryZal/manager/filters/validator/RoleValidator.java:
```diff
@@ -0,0 +1,31 @@
+        throwIfRoleIsIncorrect(role);
+    }
+
+    private void throwIfRoleIsNull(String role){
```
nazywalbym te metody bardziej valid.... albo check....

CezaryZal replied:
wydaje mi się że to trafniejsze, ewentualnie dopisać exception throwExceprionIfRolesIsNull.
Rzucamy tutaj  wyjątek, nic nie zwracamy

s235jr commented on src/main/java/com/CezaryZal/manager/modifier/entity/ContainersCreatorOfUserHcAuth.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class ContainersCreatorOfUserHcAuth {
+
+    public List<String> getRoleList(String roles){
+        if(roles.length() > 0){
+            return Arrays.asList(roles.split(","));
+        }
+        return new ArrayList<>();
+    }
+
+    public List<String> getPermissionList(String permissions){
+        if(permissions.length() > 0){
+            return Arrays.asList(permissions.split(","));
+        }
+        return new ArrayList<>();
```
Jak role będą w Set<?> a nie String to to będzie zbędne.

CezaryZal replied:
Ogółem tak, jednak czy z db czy od użytkownika role będą zawsze w Stringu

s235jr commented on src/main/java/com/CezaryZal/manager/modifier/entity/Converter.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class Converter {
```
czego converter?

CezaryZal replied:
Entity, ok dopiszę.
Nie zrobiłem bo jest w nazwie pakietu ale w sumie masz rację że nazwa klasy jest ważniejsza.
Poprawione

s235jr commented on src/main/java/com/CezaryZal/repository/UserAppRepository.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.data.repository.CrudRepository;
+
+import java.util.Optional;
+
```
@Repository

CezaryZal replied:
Ok, poprawione

s235jr commented on src/main/java/com/CezaryZal/repository/UserHcAuthRepository.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.data.repository.CrudRepository;
+
+import java.util.Optional;
+
```
@Repository

CezaryZal replied:
Racja, poprawione

CezaryZal committed:
Update by comment from pull #2 (GitHub)

Add @Repository annotation in repository package;
Add @Autowired annotation in LoginService.java;
Move PasswordEncoder bean to SecurityConfig.java and delete Bcrypt class;
Add final in UserAppPrincipal.java, UserAppPrincipalDetailsService.java,
Create InvalidPasswordException.java;
Catch AccountNotFoundException and InvalidPasswordException in RestResponseEntityExceptionHandler.java and throw another String;
Rename class to ConverterEntity.java from Converter.java in manager.modifier.entity;
Change FormValidator and FormUser classes to abstract classes;
Create ROLES enum and use him in RoleValidator.java;
Change type method to static in HeadersByToken class and HttpEntityByBodyAndToken;
Delete NewUser class;
Create UserInHcAppCreator class;

s235jr commented on src/main/java/com/CezaryZal/config/SecurityConfig.java:
```diff
@@ -0,0 +1,60 @@
+@Configuration
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    private UserAppPrincipalDetailsService detailsService;
```
final

s235jr commented on src/main/java/com/CezaryZal/config/SecurityConfig.java:
```diff
@@ -0,0 +1,60 @@
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    private UserAppPrincipalDetailsService detailsService;
+
```
autowired

CezaryZal committed:
Change project name to Server Resource

CezaryZal committed:
Add project description in README.md
